### PR TITLE
[DMS-407] Add ApiClients

### PR DESCRIPTION
**Pull request by simpat-adam**

### CODE REVIEW

File: src/config/backend/EdFi.DmsConfigurationService.Backend.Postgresql/Repositories/VendorRepository.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+                return new VendorInsertResult.Success(id);
+            }
+            catch (Exception ex)
+            {
+                await transaction.RollbackAsync();
+                return new VendorInsertResult.FailureUnknown(ex.Message);
+            }
+        }
+
+        public async Task<VendorQueryResult> QueryVendor(PagingQuery query)
```
Comment (simpat-adam) [inline src/config/backend/EdFi.DmsConfigurationService.Backend.Postgresql/Repositories/VendorRepository.cs:57]: Not using `PathingQuery` for now but will in DMS-410

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs [hunk 2]
```diff
@@ -3,12 +3,199 @@
...
+using EdFi.DmsConfigurationService.DataModel.Application;
+using EdFi.DmsConfigurationService.Frontend.AspNetCore.Infrastructure;
 using EdFi.DmsConfigurationService.Frontend.AspNetCore.Model.Validator;
+using FluentValidation;
+using FluentValidation.Results;
+using Microsoft.AspNetCore.Mvc;
 
 namespace EdFi.DmsConfigurationService.Frontend.AspNetCore.Modules;
 
-public class ApplicationModule : BaseModule<Application, ApplicationValidator>
```
Comment (simpat-adam) [inline src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:11]: Refactored away from generic "Entity/Validator" modules. Not enough "generic" code between the few entities in this application.

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs [hunk 3]
```diff
@@ -3,12 +3,199 @@
...
+        [FromServices] HttpContext httpContext,
+        [FromServices] IApplicationRepository applicationRepository,
+        [FromServices] IClientRepository clientRepository,
+        [FromServices] ILogger<ApplicationModule> logger
+    )
+    {
+        logger.LogDebug("Entering UpsertApplication");
+        await validator.GuardAsync(command);
+
+        Guid clientId = Guid.NewGuid();
```
Comment (simpat-adam) [inline src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:40]: The main reason for the ticket. Need to create Keycloak clients when application(s) are created. See also delete. For now behavior is similar to AdminApi where one client is created per Application, however the data model would support many clients per application if needed.

File: src/config/datamodel/EdFi.DmsConfigurationService.DataModel/Application/ApplicationInsertCommand.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+namespace EdFi.DmsConfigurationService.DataModel.Application
+{
+    public class ApplicationInsertCommand
```
Comment (CSR2017) [inline src/config/datamodel/EdFi.DmsConfigurationService.DataModel/Application/ApplicationInsertCommand.cs:8]: We have ApplicationInsertResponse, should we consider renaming to ApplicationInsertRequest ?

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs [hunk 5]
```diff
@@ -3,12 +3,199 @@
...
+        endpoints.MapDelete($"/v2/applications/{{id}}", Delete).RequireAuthorizationWithPolicy();
+    }
+
+    private static async Task<IResult> InsertApplication(
+        [FromServices] ApplicationInsertCommandValidator validator,
+        [FromBody] ApplicationInsertCommand command,
+        [FromServices] HttpContext httpContext,
+        [FromServices] IApplicationRepository applicationRepository,
+        [FromServices] IClientRepository clientRepository,
+        [FromServices] ILogger<ApplicationModule> logger
```
Comment (CSR2017) [inline src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:34]: We don't need to explicitly mention the [FromServices] and [FromBody]

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs [hunk 6]
```diff
@@ -3,12 +3,199 @@
...
+
+        Guid clientId = Guid.NewGuid();
+        string clientSecret = Guid.NewGuid().ToString();
+        if (
+            await clientRepository.CreateClientAsync(
+                clientId.ToString(),
+                clientSecret,
+                command.ApplicationName
+            )
+        )
```
Comment (CSR2017) [inline src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:48]: We may need explicitly handle the error from the Keycloak or backend services. Currently, system throws 500 internal server error.

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs [hunk 7]
```diff
@@ -3,12 +3,199 @@
...
+        switch (apiClientsResult)
+        {
+            case ApplicationApiClientsResult.Success success:
+                foreach (var clientUuid in success.ClientUuids)
+                {
+                    try
+                    {
+                        logger.LogInformation("Deleting client {clientUuid}", clientUuid);
+                        bool deleted = await clientRepository.DeleteClientAsync(clientUuid.ToString());
+                        if (!deleted)
```
Comment (CSR2017) [inline src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:162]: While deleting, I see we are deleting the apiclient from the Keycloak. Should we delete the record from the dmscs.ApiClient table as well ?

File: src/config/datamodel/EdFi.DmsConfigurationService.DataModel/Application/ApplicationInsertCommand.cs (hunk 4 above)
Comment (simpat-adam) [reply src/config/datamodel/EdFi.DmsConfigurationService.DataModel/Application/ApplicationInsertCommand.cs:8]: I was going for the CQRS naming pattern where broadly "C" stands for Commands to modify the state of things and "Q" is query.

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs (hunk 6 above)
Comment (simpat-adam) [reply src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:48]: Johnny is working on ticket DMS-389 for this purpose. I left for him.

File: src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs (hunk 7 above)
Comment (simpat-adam) [reply src/config/frontend/EdFi.DmsConfigurationService.Frontend.AspNetCore/Modules/ApplicationModule.cs:162]: Delete cascades are on at the database level, so they will be deleted.

**Review by CSR2017 — APPROVED** [review]

**Pull request merged**